source/features/what-changed-since-previous-release.tsx

```diff
@@ -0,0 +1,127 @@
+	const commitIdSelector = '.commit > ul a[href*="/commit/"], .release > div:first-child > ul a[href*="/commit/"]';
+
+	const allTags = extractValuesFromPathname(select.all<HTMLAnchorElement>(tagSelector), tagRegExp);
+	const allCommitIdsAnchor = select.all<HTMLAnchorElement>(commitIdSelector);
+	const allCommitIds = extractValuesFromPathname(allCommitIdsAnchor, commitRegExp);
+
+	const nextPageLink = select<HTMLAnchorElement>('.pagination a:last-child');
+	const nextPage = nextPageLink ? await fetchDom(nextPageLink.href) : await getNextPageForSinglePageTag(allTags[0]);
+
+	if (nextPage) {
+		allTags.push(...extractValuesFromPathname(select.all<HTMLAnchorElement>(tagSelector, nextPage), tagRegExp));
+		allCommitIds.push(...extractValuesFromPathname(select.all<HTMLAnchorElement>(commitIdSelector, nextPage), commitRegExp));
+	}
+
+	for (let index = 0; index < allCommitIdsAnchor.length; index++) {
+		const previousTagIndex = getPreviousTagIndex(index + 1, allCommitIds[index], allCommitIds, allTags[index], allTags);
+
+		if (previousTagIndex !== -1) {
+			allCommitIdsAnchor[index].parentElement!.parentElement!.append(getTagComparisonLink(allTags[previousTagIndex], allTags[index]));
+		}
+	}
+}
+
+const extractValuesFromPathname = (anchors: HTMLAnchorElement[], regexp: RegExp): string[] => {
+	return anchors.map((anchor: HTMLAnchorElement): string => {
+		const [, value] = anchor.pathname.match(regexp)!;
+
+		return decodeURIComponent(value);
+	});
+};
+
+const getNextPageForSinglePageTag = async (tag: string): Promise<void | DocumentFragment> => {
+	if (!isSingleTagPage()) {
+		return;
+	}
+
+	const {ownerName, repoName} = getOwnerAndRepo();
+
+	return fetchDom(`/${ownerName}/${repoName}/releases?after=${tag}`);
+};
+
+const getPreviousTagIndex = (startIndex: number, commitId: string, allCommitIds: string[], tag: string, allTags: string[]): number => {
+	let index = -1;
+
+	for (let i = startIndex; i < allCommitIds.length; i++) {
+		if (allCommitIds[i] === commitId) {
+			continue;
+		}
+
+		if (isOfSameHierarchy(allTags[i], tag) && isSameTagNameSpace(allTags[i], tag)) {
+			return i;
+		}
+
+		if (index === -1) {
+			index = i;
+		}
+	}
+
+	return index;
+};
+
+const isOfSameHierarchy = (tag1: string, tag2: string): boolean => {
+	const hierarchyRegex = /.*(\/)/;
+
+	const [tagOneHierarchy = ''] = tag1.match(hierarchyRegex) || [];
+	const [tagTwoHierarchy = ''] = tag2.match(hierarchyRegex) || [];
+
+	if ((!tagOneHierarchy && tagTwoHierarchy) || (tagOneHierarchy && !tagTwoHierarchy)) {
+		return false;
+	}
+
+	if (!tagOneHierarchy && !tagTwoHierarchy) {
+		return true;
+	}
+
+	return tagOneHierarchy === tagTwoHierarchy;
+};
+
+const isSameTagNameSpace = (tag1: string, tag2: string): boolean => {
+	const namespaceRegex = /(.*)@(.)*/;
+
+	const [, tagOneNameSpace = ''] = tag1.match(namespaceRegex) || [];
+	const [, tagTwoNameSpace = ''] = tag2.match(namespaceRegex) || [];
+
+	return tagOneNameSpace === tagTwoNameSpace;
+};
```

Comment:
If I’m understanding this correctly, hierarchy and namespace are actually part of the same thing. What we’re looking for here is SOMETHING@version. It does not matter what the shape of SOMETHING is, you just need to make sure that SOMETHING is the same for both links, if it exists. Example:

```js
function doesNamespaceMatch() {
  const [namespace1] = tag1.split(/@[^@]+$|^/); // I think this regex will always keep the version at index 1 of the array
  const [namespace2] = tag2.split(/@[^@]+$|^/);
  return namespace1 === namespace2;
}
```

Reply:
Much better. 🎉

Reply:
I'm working on the regex, it doesn't what I had hoped

Reply:
Ok, now it sort of does. It relies on the assumption that there's no such thing as `v1.2@v1.2` tag, so even if in a namespace-less tag the version is at the index `0`, it won't affect the comparison:

<img width="311" alt="" src="https://user-images.githubusercontent.com/1402241/57007585-50ae6000-6c1c-11e9-87a7-a49dcccb3e70.png">

Reply:
Hey @bfred-it, How did you know I was doing the same? 🤔 Simplifying the regex. I just changed regex from `/@[^@]+$|^/` to `/@[^@]+$/`. You spyin' on me? 😂

Reply:
I had added `|^` thinking that it would split `v1.2` into `['', 'v1.2`]`, but that only broke the `@user/module@version`:

<img width="313" alt="Screenshot 2019-05-01 at 14 26 35" src="https://user-images.githubusercontent.com/1402241/57007697-34f78980-6c1d-11e9-888e-c637c79e16bf.png">